background.js

```diff
@@ -11,31 +19,94 @@ function last(array) {
     return array[array.length - 1];
 }
 
-// Open the origin url of the given tab.
-function openTabOrigin(tab) {
-    const id = tab.id.toString();
-    api.storage.local.get(id, function(result) {
-        const result_stack = result[id] || [];
-        if (last(result_stack)) {
-            api.tabs.query({url: last(result_stack)}, function(matches) {
-                if (matches.length > 0) {
-                    api.tabs.update(matches[0].id, {active: true});
-                    api.windows.update(matches[0].windowId, {focused: true});
-                } else {
-                    const dest = last(result_stack);
-                    api.tabs.create({url: dest, index: tab.index}, function(newtab) {
-                        // We don't want to set the last tab to the one we just came
-                        // from (this one), instead we want to inherit the parent tab
-                        // stack so we can keep going all the way back.
-                        api.storage.local.set({[newtab.id.toString()]: result_stack.slice(0, -1)});
-                    });
-                }
-            });
-        } else {
-            console.log("Could not find origin for tab", id);
-            api.browserAction.setBadgeText({text: "N/A", tabId: tab.id});
-        }
-    });
+api.storage.local.onChanged.addListener(changes => {
+    const actionType = changes[actionTypeKey]
+    // looking only for "action type" changes
+    if (actionType) {
+        currentActionType = actionType.newValue
+    }
+})
+
+/** 
+ * React on extension button click and execute
+ * current action (open tab, show popup, etc.)
+ */
+function extensionButtonClick(tab) {
+
+    lastTabOriginData = {}
+
+    const tabId = tab.id.toString();
+
+    // TODO since bg script is not persistent
+    // it is possible that `currentActionType` will be `undefined`
```

Comment:
do you happen to know what conditions trigger this?

Reply:
In which circumstance bg script is terminated? I don't know. If you are talking about `currentActionType` - I don't remember, this comment looks to me as a complete mystery right now :sweat_smile: . Maybe I got a couple of null-ref errors. I assume the comment it is not relevant anymore,